src/utils/exchange_transfer_simulator.ts

```diff
@@ -0,0 +1,134 @@
+        if (_.isUndefined(this._balance[tokenAddress]) || _.isUndefined(this._balance[tokenAddress][userAddress])) {
+            const balance = await this._token.getBalanceAsync(tokenAddress, userAddress);
+            this.setBalance(tokenAddress, userAddress, balance);
+        }
+        return this._balance[tokenAddress][userAddress];
+    }
+    protected setBalance(tokenAddress: string, userAddress: string, balance: BigNumber.BigNumber): void {
+        if (_.isUndefined(this._balance[tokenAddress])) {
+            this._balance[tokenAddress] = {};
+        }
+        this._balance[tokenAddress][userAddress] = balance;
+    }
+    protected async getProxyAllowanceAsync(tokenAddress: string, userAddress: string): Promise<BigNumber.BigNumber> {
+        if (_.isUndefined(this._proxyAllowance[tokenAddress]) ||
+            _.isUndefined(this._proxyAllowance[tokenAddress][userAddress])) {
+            const proxyAllowance = await this._token.getProxyAllowanceAsync(tokenAddress, userAddress);
+            this.setProxyAllowance(tokenAddress, userAddress, proxyAllowance);
+        }
+        return this._proxyAllowance[tokenAddress][userAddress];
+    }
+    protected setProxyAllowance(tokenAddress: string, userAddress: string, proxyAllowance: BigNumber.BigNumber): void {
+        if (_.isUndefined(this._proxyAllowance[tokenAddress])) {
+            this._proxyAllowance[tokenAddress] = {};
+        }
+        this._proxyAllowance[tokenAddress][userAddress] = proxyAllowance;
+    }
+}
+
+export class ExchangeTransferSimulator extends BalanceAndProxyAllowanceLazyStore {
+    constructor(token: TokenWrapper) {
+        super(token);
+    }
+    /**
+     * Simulates transferFrom call performed by a proxy
+     * @param  tokenAddress      Address of a token to be transfered
+     * @param  from              Owner of the transfered tokens
+     * @param  to                Recipient of the transfered tokens
+     * @param  amountInBaseUnits The amount of tokens being transfered
+     * @param  tradeSide         Is Maker/Taker transferring
+     * @param  transferType      Is it a fee payment or a value transfer
+     */
+    public async transferFromAsync(tokenAddress: string, from: string, to: string,
+                                   amountInBaseUnits: BigNumber.BigNumber, tradeSide: TradeSide,
+                                   transferType: TransferType): Promise<void> {
+        const balance = await this.getBalanceAsync(tokenAddress, from);
+        const proxyAllowance = await this.getProxyAllowanceAsync(tokenAddress, from);
+        if (balance.lessThan(amountInBaseUnits)) {
+            this.throwValidationError(FailureReason.Balance, tradeSide, transferType);
+        }
+        if (proxyAllowance.lessThan(amountInBaseUnits)) {
+            this.throwValidationError(FailureReason.ProxyAllowance, tradeSide, transferType);
+        }
```

Comment:
Let's prioritize throwing allowance errors before balance errors since the user flow would be for the user to first set an allowance and then transfer.